gammapy/scripts/jupyter.py

```diff
@@ -0,0 +1,157 @@
+                    has_semicolon = fmt.endswith(';')
+                    fmt = format_str(src_contents=fmt,
+                                     line_length=79).rstrip()
+                    if has_semicolon:
+                        fmt += ';'
+                except Exception as ex:
+                    logging.info(ex)
+                fmt = fmt.replace("###-MAGIC COMMAND-", "")
+            cell['source'] = fmt
+
+        if rawnb:
+            nbformat.write(rawnb, str(path))
+            log.info('Jupyter notebook {} painted in black.'.format(str(path)))
+
+
+@click.command(name='stripout')
+@click.pass_context
+def cli_jupyter_stripout(ctx):
+    """Strip output cells."""
+
+    for path in build_nblist(ctx):
+        rawnb = read_notebook(path)
+
+        for cell in rawnb.cells:
+            if cell['cell_type'] == 'code':
+                cell['outputs'] = []
+
+        if rawnb:
+            nbformat.write(rawnb, str(path))
+            log.info('Jupyter notebook {} output stripped.'.format(str(path)))
+
+
+@click.command(name='execute')
+@click.pass_context
+def cli_jupyter_execute(ctx):
+    """Execute jupyter notebook."""
+
+    for path in build_nblist(ctx):
+        run_notebook(path)
+
+
+@click.command(name='test')
+@click.pass_context
+def cli_jupyter_test(ctx):
+    """Check if Jupyter notebook is broken."""
+
+    for path in build_nblist(ctx):
+        run_notebook(path)
+        rawnb = read_notebook(path)
+
+        if rawnb:
+            report = ""
+            passed = True
+            log.info("   ... Testing {}".format(str(path)))
+            for cell in rawnb.cells:
+                if 'outputs' in cell.keys():
+                    for output in cell['outputs']:
+                        if output['output_type'] == 'error':
+                            passed = False
+                            traceitems = ["--TRACEBACK: "]
+                            for o in output['traceback']:
+                                traceitems.append("{}".format(o))
+                            traceback = "\n".join(traceitems)
+                            infos = "\n\n{} in cell [{}]\n\n" \
+                                    "--SOURCE CODE: \n{}\n\n".format(
+                                        output['ename'],
+                                        cell['execution_count'],
+                                        cell['source']
+                                    )
+                            report = report + infos + traceback
+            if passed:
+                log.info("   ... {} Passed".format(str(path)))
+            else:
+                log.info("   ... {} Failed".format(str(path)))
+                log.info(report)
+
+
+def tag_magics(cellcode):
+    """Comment magic commands when formatting cells."""
+
+    lines = cellcode.splitlines(False)
+    for line in lines:
+        if line.startswith("%") or line.startswith("!"):
+            magic_line = "###-MAGIC COMMAND-" + line
```

Comment:
This `tag_magics` function appears very far from where it's called, making the code overall not so easy to read. Also the magic string "###-MAGIC COMMAND-" shouldn't be duplicated in two places.

If blackening a notebook needs more than one function, i.e. a second helper function makes sense, I would suggest to group the two functions in a class. The task of the class would be to process one notebook., the loop over notebooks would remain in the click cli function.